MapboxGLAndroidSDK/src/cpp/style/sources/source.cpp

```diff
@@ -50,205 +55,263 @@ static std::unique_ptr<Source> createSourcePeer(jni::JNIEnv& env,
     }
+    return coreSource.peer.get<Source*>()->javaPeer;
+}
 
-    const jni::Object<Source>& Source::peerForCoreSource(jni::JNIEnv& env, mbgl::style::Source& coreSource) {
-        if (!coreSource.peer.has_value()) {
-            coreSource.peer = createSourcePeer(env, coreSource, nullptr);
-        }
-        return coreSource.peer.get<std::unique_ptr<Source>>()->javaPeer;
+const jni::Object<Source> &
+Source::peerForCoreSource(jni::JNIEnv &env, mbgl::style::Source &coreSource) {
+    if (!coreSource.peer.has_value()) {
+        coreSource.peer = createSourcePeer(env, coreSource, nullptr).release();
     }
+    return coreSource.peer.get<Source*>()->javaPeer;
+}
 
-    Source::Source(jni::JNIEnv& env,
-                   mbgl::style::Source& coreSource,
-                   const jni::Object<Source>& obj,
-                   AndroidRendererFrontend* frontend)
-        : source(coreSource), javaPeer(jni::NewGlobal(env, obj)), rendererFrontend(frontend) {}
+Source::Source(jni::JNIEnv &env,
+               mbgl::style::Source &coreSource,
+               const jni::Object<Source> &obj,
+               AndroidRendererFrontend *frontend)
+        : source(coreSource.makeWeakPtr()), javaPeer(jni::NewGlobal(env, obj)), rendererFrontend(frontend) {}
 
-    Source::Source(jni::JNIEnv&, std::unique_ptr<mbgl::style::Source> coreSource)
-        : ownedSource(std::move(coreSource))
-        , source(*ownedSource) {
+Source::Source(jni::JNIEnv &, std::unique_ptr<mbgl::style::Source> coreSource)
+        : ownedSource(std::move(coreSource)), source(ownedSource->makeWeakPtr()) {
+}
+
+Source::~Source() {
+    if (ownedSource) {
+        ownedSource.reset();
+        ownedSource.release();
     }
 
-    Source::~Source() {
-        if (ownedSource) {
-            ownedSource.reset();
-            ownedSource.release();
-        }
-        // Before being added to a map, the Java peer owns this C++ peer and cleans
-        //  up after itself correctly through the jni native peer bindings.
-        // After being added to the map, the ownership is flipped and the C++ peer has a strong reference
-        //  to it's Java peer, preventing the Java peer from being GC'ed.
-        //  In this case, the core source initiates the destruction, which requires releasing the Java peer,
-        //  while also resetting it's nativePtr to 0 to prevent the subsequent GC of the Java peer from
-        //  re-entering this dtor.
-        if (ownedSource.get() == nullptr && javaPeer.get() != nullptr) {
-            // Manually clear the java peer
-            android::UniqueEnv env = android::AttachEnv();
-            static auto& javaClass = jni::Class<Source>::Singleton(*env);
-            static auto nativePtrField = javaClass.GetField<jlong>(*env, "nativePtr");
-            javaPeer.Set(*env, nativePtrField, (jlong) 0);
-            javaPeer.reset();
+    if (ownedSource.get() == nullptr && javaPeer.get() != nullptr) {
```

Comment:
`ownedSource.get()` must be nullptr, as it is released on line https://github.com/mapbox/mapbox-gl-native-android/pull/432/files#diff-2c2b9946805e54eb4377357eca5763c3R87

Reply:
Right.